File: Algorithms/src/3003. Maximize the Number of Partitions After Operations/Code/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string s;
    int k;
    unordered_map<long long, int> memo;

    int dp(int i, long long mask, bool canChange) {
        if (i == s.size()) return 0;
        long long key = ((long long)i << 30) | (mask << 1) | canChange;
        if (memo.count(key)) return memo[key];

        int bit = s[i] - 'a';
        long long newMask = mask | (1LL << bit);
        int res = 0;

        // If current partition exceeds k distinct chars, we start new
        if (__builtin_popcountll(newMask) > k)
            res = 1 + dp(i + 1, 1LL << bit, canChange);
        else
            res = dp(i + 1, newMask, canChange);

        // Try changing this character (if still available)
        if (canChange) {
            for (int j = 0; j < 26; ++j) {
                long long changeMask = mask | (1LL << j);
                if (__builtin_popcountll(changeMask) > k)
                    res = max(res, 1 + dp(i + 1, 1LL << j, false));
                else
                    res = max(res, dp(i + 1, changeMask, false));
            }
        }

        return memo[key] = res;
    }

    int maxPartitionsAfterOperations(string s, int k) {
        this->s = s;
        this->k = k;
        memo.clear();
        return dp(0, 0, true) + 1;
    }
};

File: Algorithms/src/3003. Maximize the Number of Partitions After Operations/Code/solution.go
package main

func maxPartitionsAfterOperations(s string, k int) int {
	n := len(s)
	cache := make(map[[3]int]int)

	var bitCount = func(x int) int {
		count := 0
		for x > 0 {
			count += x & 1
			x >>= 1
		}
		return count
	}

	var dp func(int, int, int) int
	dp = func(i int, canChange int, mask int) int {
		if i == n {
			return 0
		}
		key := [3]int{i, canChange, mask}
		if val, ok := cache[key]; ok {
			return val
		}
		bit := int(s[i] - 'a')
		newMask := mask | (1 << bit)
		res := 0

		if bitCount(newMask) > k {
			res = 1 + dp(i+1, canChange, 1<<bit)
		} else {
			res = dp(i+1, canChange, newMask)
		}

		if canChange == 1 {
			for j := 0; j < 26; j++ {
				changed := mask | (1 << j)
				if bitCount(changed) > k {
					res = max(res, 1+dp(i+1, 0, 1<<j))
				} else {
					res = max(res, dp(i+1, 0, changed))
				}
			}
		}
		cache[key] = res
		return res
	}

	return dp(0, 1, 0) + 1
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
